### DISCUSSION TITLE
How to pass node text to a shell command as input and get back the output?

### DISCUSSION THREAD

**Discussion by ysaereve** [freeplane/freeplane#504] [Help] [answered]
I am trying to pass node text to a shell command as input and get back the output. With the information from Stack Overflow discussion, I am able to capture the output. The problem that I have is I am not able to use either StringBufferWriter or StringBuilderWriter in my script to send string as standard input. Could you share with me your suggestion?

### Reference
Groovy executing shell commands - Stack Overflow
https://stackoverflow.com/questions/159148/groovy-executing-shell-commands

Process (Groovy JDK)
http://docs.groovy-lang.org/docs/groovy-1.7.2/html/groovy-jdk/java/lang/Process.html

### My Code
```
def /*n*/ nRunCommand(n) {
	def sout = new StringBuilder(), serr = new StringBuilder()
 	def sin = new StringBuilder(), win = new StringBufferWriter(sin)
	def proc = 'nl -ba'.execute()
	proc.consumeProcessOutput(sout, serr)
	/* using withWriter? and how? */
	proc.waitForOrKill(1000)
	n.text=sout.toString().trim()+" "+n.text
}
```
### Errors
```
STDOUT: Jul 17, 2022 1:11:01 PM org.freeplane.core.util.LogUtils warn
WARNING: error executing script /Users/lieric/.freeplane/1.10.x/scripts/ScriptTestDriveScript.groovy - giving up
java.security.PrivilegedActionException: org.codehaus.groovy.control.MultipleCompilationErrorsException: startup failed:
~/.freeplane/1.10.x/scripts/ScriptTestDriveScript.groovy: 47: unable to resolve class StringBuilderWriter
 @ line 47, column 40.
    	def sin = new StringBuilder(), win = new StringBuilderWriter(sin)
                                          ^
1 error
```

**Comment by macmarrum** [comment, marked as answer]
This is what works for me.
```groovy
def proc = ['python', '-c', 'import sys; sys.stdout.write(sys.stdin.read().replace("a", "X"))'].execute()
proc.out.withWriter("UTF-8") { it.write(node.text) }
node.text = proc.text
```
Please note that the code uses Groovy enhancements to create a process -- https://groovy-lang.org/groovy-dev-kit.html#process-management
The example executes `python` (must be in PATH), which processes the input, replacing `a` with `X`.
`node.text` is passed as the input to the process. NB input is exposed as `proc.out`, as explained in the Groovy docs:
> It is worth noting that `in` corresponds to an input stream to the standard output of the command. `out` will refer to a stream where you can send data to the process (its standard input).

**Reply by ysaereve** [reply]
It is concise and works great! 👍 Thank you so much for the help and additional notes.

**Comment by dpolivaev** [comment]
The compiler says that there is no class `StringBuilderWriter` and it is true.

**Reply by ysaereve** [reply]
Thank you for sharing this information, and thank you for building this wonderful tool - Freeplane.
